### PR TITLE
Issue 1118: refine HandlerBase to let only ConsumerImpl and ProducerImpl have client-cnx

### PR DESCRIPTION
**Pull request by zhaijack**
### Motivation

 class HandlerBase contains 2 parts:
- handler state;  this is need for all handler.
- client_cnx connection handling; this part only useful in ConsumerImpl and ProducerImpl, while not need for PartitionedConsumer, PartitionedProducer, TopicsConsumer.

This change would like to refine HandlerBase into 2 parts, to make client_cnx connection handling only served in ConsumerImpl and ProducerImpl

### Modifications

Split HandlerBase into 2 class: 
- HandlerState: handling the state of every consumer/producer;
- ConnectionHandler: handling the client_cnx connection for ConsumerImpl and ProducerImpl.

### Result
no api changes.

### CODE REVIEW

**Comment by zhaijack** [comment]
retest this please

**Comment by zhaijack** [comment]
retest this please

**Comment by zhaijack** [comment]
retest this please

**Comment by zhaijack** [comment]
retest this please

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageIdTest.java [hunk 1]
```diff
@@ -454,11 +454,6 @@ public void testCorruptMessageRemove() throws Exception {
         }
 
         // 5. Verify
-
-        // (5.1) Verify: producer's recoverChecksumError and updateChecksum invoked
-        verify(producer, times(1)).recoverChecksumError(any(), anyLong());
-        verify(producer, times(1)).verifyLocalBufferIsNotCorrupted(any());
-
         /**
          * (5.3) verify: ProducerImpl.verifyLocalBufferIsNotCorrupted() => validates if message is corrupt
```
Comment (sijie) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/MessageIdTest.java]: update "(5.3)"?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBase.java [hunk 2]
```diff
@@ -30,16 +30,16 @@
...
     protected final CompletableFuture<Producer<T>> producerCreatedFuture;
     protected final ProducerConfigurationData conf;
     protected final Schema<T> schema;
 
     protected ProducerBase(PulsarClientImpl client, String topic, ProducerConfigurationData conf,
             CompletableFuture<Producer<T>> producerCreatedFuture, Schema<T> schema) {
-        super(client, topic, new Backoff(100, TimeUnit.MILLISECONDS, 60, TimeUnit.SECONDS,
-                Math.max(100, conf.getSendTimeoutMs() - 100), TimeUnit.MILLISECONDS));
+        super(client, topic);
+        //new Backoff(100, TimeUnit.MILLISECONDS, 60, TimeUnit.SECONDS, Math.max(100, conf.getSendTimeoutMs() - 100), TimeUnit.MILLISECONDS));
```
Comment (sijie) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerBase.java]: nit: do we still need this?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/HandlerState.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.client.impl;
+
+import java.util.concurrent.atomic.AtomicReferenceFieldUpdater;
+import java.util.function.UnaryOperator;
+
+public abstract class HandlerState {
```
Comment (sijie) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/HandlerState.java]: nit: we don't need to make it `public`.

**Comment by zhaijack** [comment]
Thanks @sijie, changed following your comments.

**Pull request merged**